internal/argocd/revision.go

```diff
@@ -2,107 +2,224 @@ package argocd
+	appLogger.Debug("Getting desired revision for application", "app", app, "spec", app.Spec)
+
 	// Note that frght was provided as an argument instead of being plucked
 	// directly from stage.Status, because this gives us the flexibility to use
 	// this function for finding the revision to sync to either in the context of
 	// a health check (current freight) or in the context of a promotion (new
 	// freight).
+
+	// An application may have one or more sources.
+	if !app.IsMultisource() {
+		if app.Spec.Source == nil {
+			err := errors.New("Single-source application source is nil")
+			appLogger.Error(err, "Cannot determine desired revision for application, bailing.")
+			return NO_REVISIONS, nil
+		}
+
+		appLogger.Debug("Application source is not nil, checking.", "source", app.Spec.Source)
+		desiredRevision, err := getRevisionFromSource(ctx, cl, stage, update, app.Spec.Source, frght)
+
+		if err != nil {
+			return NO_REVISIONS, err
+		}
+
+		if desiredRevision != "" {
+			return []string{desiredRevision}, nil
+		}
+	}
+
+	// An application may have more than one source that points to the same Git repository,
+	// eg. a Helm and a vanilla manifest source.
+	// In that situation it doesn't matter which source's target revision is returned as ArgoCD does not support
+	// different target revisions for sources targeting the same repository.
+	var revisions []string
+	for i := range app.Spec.Sources {
+		s := &app.Spec.Sources[i]
+		desiredRevision, err := getRevisionFromSource(ctx, cl, stage, update, s, frght)
+
+		if err != nil {
+			return NO_REVISIONS, err
+		}
+
+		if desiredRevision != "" {
+			appLogger.Trace("Found desired revision for application source.", "revision", desiredRevision, "source", s)
+			revisions = append(revisions, desiredRevision)
+		}
+	}
+	if len(revisions) > 0 {
+		appLogger.Debug("Found desired revision(s) for multi-source application.", "revisions", revisions)
+		return revisions, nil
+	}
+
+	// If we end up here, no desired revision was found from any of the possible sources.
+	appLogger.Debug("Could not determine desired revision for application from any sources.")
+	return NO_REVISIONS, nil
+}
+
+func getRevisionFromSource(
+	ctx context.Context,
+	cl client.Client,
+	stage *kargoapi.Stage,
+	update *kargoapi.ArgoCDAppUpdate,
+	s *argocd.ApplicationSource,
+	frght []kargoapi.FreightReference) (string, error) {
+
+	logger := logging.LoggerFromContext(ctx)
+	sourceLogger := logger.WithValues("source", s)
+	sourceLogger.Debug("Getting desired revision for application source", "source", s)
+
+	// If there is a source update that targets app.Spec.Source, it might
+	// have its own ideas about the desired revision.
+	desiredOrigin := getDesiredOriginForSource(ctx, stage, update, s)
+
 	switch {
-	case app == nil || app.Spec.Source == nil:
-		// Without an Application, we can't determine the desired revision.
-		return "", nil
-	case app.Spec.Source.Chart != "":
-		// This source points to a Helm chart.
```

Comment:
I would maybe not lose this comment.

Reply:
My code starts with logging the same, is this necessary?
```go
	case s.Chart != "":
		// NB: This has to go first, as the repository URL can also point to
		//     a Helm repository.
		sourceLogger.Debug("Application source is a Helm chart")
```

Reply:
I think there's a separate convo to be had about logging. @hiddeco and I both felt the amount of new logging introduced in this PR may be excessive. I don't want to get hung up on that right now though, as that's something that can be discussed later after we get this PR working.

When reviewing this PR, purely to aid my own focus, I reviewed it locally with log-related changes removed to reduce the diffs to just those that were functionally relevant. It was from that myopic POV sans the log message that I found myself wishing this comment had remained.

I agree the log message and the comment are redundant and we only need one or the other. Leave it be for now and we can circle back on this later after the important stuff is dealt with.